tools/2025-03-workflow-changes.md

```diff
@@ -0,0 +1,260 @@
+* Increase the reliability of workflows.
+* Reduce time dedicated to addressing workflow failures by the on-call engineer.
+
+### Non-goals
+
+* This design does not include details on how each Radius workflow would be modified to align with the patterns defined here.
+* Not every workflow must be updated to the principles defined in this design. We will prioritize refactoring workflows based on the level of effort each workflow is requiring from the team when we need to make changes or when the workflows fail and need attention.
+* Replacement of security tokens, e.g. PATs, with GitHub applications.
+
+### User scenarios
+
+#### User story: Forks
+
+As a Radius developer, I can fork a repo and run any workflow on the fork with minimal setup, so that I can quickly test changes and confidently submit a PR for those changes.
+
+#### User story: Local debugging
+
+As a Radius developer, I can debug and test workflow logic on my machine, without having to commit the workflow changes to a repository, so that I can efficiently finish my work and ensure high quality code in the workflows.
+
+#### User story: Security context
+
+As a Radius developer, I can establish a security context that works on my machine or within a GitHub workflow, so that I can easily test automation developed locally, and without code changes I can run that automation within a GitHub workflow.
+
+#### User story: Review workflow process
+
+As a PR reviewer, I can checkout the PR to my own fork and test workflows there so that I can ensure the quality of the workflow.
+
+### Developer Experience
+
+Developers should be able to follow a standard development process when modifying GitHub workflows. Reviewers can test by checking out the PR to their own fork.
+
+```mermaid
+---
+title: Developer and PR Reviewer Experience
+---
+flowchart LR
+    subgraph Developer
+        direction TB
+        fork["Fork the repo"]
+        clone["Clone and branch"]
+        code["Modify code"]
+        testlocal["Test on dev machine"]
+        pushtofork["Test on GitHub fork"]
+        pr["Create PR"]
+
+        fork --> clone
+        clone --> code
+        code --> testlocal
+        testlocal --> pushtofork
+        pushtofork --> pr
+    end
+
+    subgraph Reviewer
+        direction TB
+        reviewpr["Checkout PR to Fork"]
+        testonfork["Test on Fork"]
+        prfeedback["Provide Feedback"]
+        reviewpr --> testonfork
+        testonfork --> prfeedback
+    end
+
+    Developer --> Reviewer
+    %%Reviewer --> Developer
+```
+
+_Fig. 1: The developer and PR reviewer steps_
+
+## Design
+
+### High Level Workflow Design Principles
+
+#### Use GitHub workflows for CI/CD setup and workflow layout
+
+We will use GitHub workflows for these things:
+
+* Identity and Security
+* Test runner setup
+* Control flow, including matrix operations
+
+GitHub workflows will not contain core logic (unless the workflow is very simple).
```

Comment:
Could you add an example of "core logic" and the rationale behind this principle?

Reply:
Never mind, I see the definition for `core logic` in terms and definitions and some details on why later in this doc. I think it will help to add brief context here in this section as well.

Reply:
I expanded the definition of core logic to clarify.